pkg/ring/lifecycler.go

```diff
@@ -242,12 +252,59 @@ func (i *Lifecycler) getTokens() []uint32 {
 	return i.tokens
 }
 
+func (i *Lifecycler) flushTokensToFile() {
+	if i.cfg.TokenFileDir == "" {
+		return
+	}
+	tokenFilePath := path.Join(i.cfg.TokenFileDir, tokensFileName)
+	f, err := os.OpenFile(tokenFilePath, os.O_WRONLY|os.O_CREATE, 0666)
+	if err != nil {
+		level.Error(util.Logger).Log("msg", "error in creating token file", "err", err)
+		return
+	}
+
+	b := make([]byte, 4*len(i.tokens))
```

Comment:
I would decouple the tokens storage format from the ring lifecycler logic. I would suggest to create a dedicated `type` for the tokens with serialization and deserialization functions. I would also suggest to add a version number to the format, to keep the door open to future changes.

Reply:
If we were to have versions for tokens here, would we also need to store the version in the ring too?

Reply:
But I don't see how we will have different versions of tokens in the future.

Reply:
The comment was about two distinct things:
1. I would suggest to create a dedicated type for the tokens with serialization and deserialization functions
2. I would also suggest to add a version number to the format, to keep the door open to future changes.

I'm still keen on the comment #1. It helps to keep a separation of concerns and the code cleaner and easier to unit test.

About #2 I see your point. My point is that it's hard to foresee the future, and adding a byte with a version number at the beginning of the file looks a minimal effort compared to the case in the future we'll have to change the format (ie. add other data) and versioning will be handy.

Reply:
Will add the point 1. About point 2, I will add the type part in this PR to store in the disk but we might need to change it in the ring too which can be a follow-up PR with some migration path.

Reply:
About point 2, ask other people feedback, please. Don't want to be pushy on it. Realistically, there may not be that need and may be not worth.

Reply:
> But I don't see how we will have different versions of tokens in the future.

Not sure if that would need to be stored, but #1764 is introducing State to the tokens.

If you need extensible serialization format, perhaps just use protobuf.

Reply:
Added an interface to handle tokens https://github.com/cortexproject/cortex/pull/1750/commits/7ab02ffe583ea85795353e8db0909d68b9794378. Might be an initial iteration towards the final version, is it what is intended? I tried to have `type Token uint32` but that would have led to a lot of changes and I would like to keep that for a separate PR in itself.

Reply:
I left my feedback here:
https://github.com/cortexproject/cortex/pull/1750#pullrequestreview-311208119